File: 001-TwoSum/python/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys

PYTHON_COMMON_PATH = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../../common/python_common'))
if PYTHON_COMMON_PATH not in sys.path:
    sys.path.insert(0, PYTHON_COMMON_PATH)

import common.common as common

class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        ret = []
        n = len(nums)
        for i in range(0, n - 1):
            for j in range(i + 1, n):
                if nums[i] + nums[j] == target:
                    ret.append(i)
                    ret.append(j)
                    return ret
        return ret
    def twoSum_2(self, nums, target):
        ret = []
        n = len(nums)
        hmap = {}
        for i in range(0, n):
            hmap[nums[i]] = i
        for i in range(0, n):
            sub = target - nums[i]
            if hmap.has_key(sub) and hmap[sub] != i:
                ret.append(i)
                ret.append(hmap[sub])
                return ret
        return ret


if __name__ == '__main__':
    try:
        solution = Solution()
        ret = common.get_string_from_file("./data1.txt")
        if ret["result"] != 0:
            print ret["msg"]
            exit
        string = ret["string"]
        
        ret = common.split(string, "\n")
        if ret["result"] != 0:
            print ret["msg"]
            exit
        strs = ret["array"]
        
        ret = common.strings_to_ints(strs)
        if ret["result"] != 0:
            print ret["msg"]
            exit
        array = ret["array"]
        ret = solution.twoSum(array, 9)
        print ret
        ret = solution.twoSum_2(array, 9)
        print ret
    except Exception as ex:
        print ex

File: 001-TwoSum/c++/main.cpp
#include <iostream>
#include <vector>
#include "common/cpp_common.h"
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        int n = nums.size();
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j <= n - 1; j++) {
                if (nums[i] + nums[j] == target) {
                    result.push_back(i);
                    result.push_back(j);
                    break;
                }
            }
        }
        return result;
    }
};

class Solution2 {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> result;
        int n = nums.size();
        std::unordered_map<int, int> hmap;
        for (int i = 0; i < nums.size(); i++) {
            hmap[nums[i]] = i;
        }
        for (int i = 0; i < n - 1; i++) {
            int sub = target - nums[i];
            auto iter = hmap.find(sub);
            if (iter != hmap.end() && iter->second != i) {
                result.emplace_back(i);
                result.emplace_back(iter->second);
                break;
            }
        }
        return result;
    }
};

int main(int argc, char *argv[]) {
    std::vector<std::string> strs;
    std::string str;
    leetcode::common::get_string_from_file("./data1.txt", str);
    strs = leetcode::common::split(str,'\n');

    std::vector<int> array;
    leetcode::common::strings_to_numbers(strs, array);
    // 解法一
    Solution *solution = new Solution();
    vector<int> result = solution->twoSum(array, 9);

    for(int i = 0; i < result.size(); i++) {
        cout << result[i] << " ";
    }
    cout << endl;
    // 解法二
    Solution2 *solution2 = new Solution2();
    vector<int> result2 = solution2->twoSum(array, 9);

    for(int i = 0; i < result2.size(); i++) {
        cout << result2[i] << " ";
    }
    cout << endl;

    return 0;
}
